src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala

```diff
@@ -0,0 +1,104 @@
+import org.apache.flink.ml.common.ParameterMap
+import org.apache.flink.streaming.api.scala.DataStream
+import org.apache.flink.streaming.api.scala.function.WindowFunction
+import org.apache.flink.streaming.api.windowing.assigners.GlobalWindows
+import org.apache.flink.streaming.api.windowing.triggers.Trigger.TriggerContext
+import org.apache.flink.streaming.api.windowing.triggers.{Trigger, TriggerResult}
+import org.apache.flink.streaming.api.windowing.windows.GlobalWindow
+import org.apache.flink.util.Collector
+
+class FitTrigger[W <: GlobalWindow] extends Trigger[LassoStreamEvent, W] {
+  override def onElement(element: LassoStreamEvent, timestamp: Long, window: W, ctx: TriggerContext): TriggerResult = {
+    element match {
+      case Left(ev) => TriggerResult.FIRE
+      case Right(_) => TriggerResult.CONTINUE
+    }
+  }
+
+  override def clear(w: W, triggerContext: TriggerContext): Unit = ???
+
+  override def onProcessingTime(time: Long, window: W, ctx: TriggerContext): TriggerResult = {
+    TriggerResult.CONTINUE
+  }
+  override def onEventTime(time: Long, window: W, ctx: TriggerContext): TriggerResult = {
+    TriggerResult.CONTINUE
+  }
+}
+
+class FitWindowFunction extends WindowFunction[LassoStreamEvent, OptionLabeledVector, Long, GlobalWindow] {
+
+  def apply(key: Long, window: GlobalWindow, input: Iterable[LassoStreamEvent],
+            out: Collector[OptionLabeledVector]): Unit =
+  {
+    val inputStreamEvents = input.filter(x => x.isLeft).map(x => x.left.get)
+    val labelStreamEvent = input.filter(x => x.isRight).map(x => x.right.get).head
+
+    val poses = inputStreamEvents.toList.map(x => x.pos._2)
+    var labels = List[(Double, Double)]()
+
+    for (i <- labelStreamEvent.labels.data.indices) {
+      labels = (labelStreamEvent.poses(i), labelStreamEvent.labels(i)) :: labels
+    }
+
+    val interpolatedLabels = new FlatnessMappingAlgorithm(poses, labels).apply
+
+    for((event, index) <- inputStreamEvents.zipWithIndex){
+      val labeledEvent: OptionLabeledVector = Left((event.data.asBreeze, interpolatedLabels(index)))
+      out.collect(labeledEvent)
+    }
+  }
+}
+
+class LassoDFStreamTransformOperation[T <: LassoStreamEvent]
+    extends TransformDataStreamOperation[LassoDelayedFeedbacks, LassoStreamEvent, Option[Double]]{
+
+    override def transformDataStream(instance: LassoDelayedFeedbacks,
+                                     transformParameters: ParameterMap,
+                                     rawInput: DataStream[LassoStreamEvent]): DataStream[Option[Double]] = {
+      def selectKey(event: LassoStreamEvent): Long = {
+        event match {
+          case Left(ev) => ev.pos._1
+          case Right(ev) => ev.label
+        }
+      }
+      val processedInput = rawInput.keyBy(x => selectKey(x)).window(GlobalWindows.create())
+        .trigger(new FitTrigger).apply(new FitWindowFunction())
+
+      val output = LassoParameterServer.transformLasso(None)(processedInput, workerParallelism = 3,
```

Comment:
why do you bound the parallelism to 3? Wouldnt it be better to extract the parallelism from the execution environment?

Reply:
Moreover, there are a lot of hard coded params. This is ok at this stage, imho. However, we might need to rethink this part later on (or generalize those settings now).